src/codegen/sys/tests.rs

```diff
@@ -0,0 +1,290 @@
+use std::io::prelude::*;
+use std::io;
+use std::path::Path;
+
+use analysis::types::IsIncomplete;
+use env::Env;
+use file_saver::save_to_file;
+use library::{Type, MAIN_NAMESPACE};
+use nameutil::crate_name;
+use codegen::general;
+
+struct CType {
+    /// Name of type, as used in C.
+    name: String,
+    /// Expression describing when type is available (when defined only conditionally).
+    cfg_condition: Option<String>,
+}
+
+pub fn generate(env :&Env) {
+    let tests = env.config.target_path.join("tests");
+    let abi_c = tests.join("abi.c");
+    let abi_rs = tests.join("abi.rs");
+    let manual_h = tests.join("manual.h");
+
+    let ns = env.library.namespace(MAIN_NAMESPACE);
+    let ctypes = ns.types
+        .iter()
+        .filter_map(|t| t.as_ref())
+        .filter(|t| !t.is_incomplete(&env.library))
+        .filter_map(|t| match *t {
+            Type::Alias(_) |
+            Type::Class(_) |
+            Type::Record(_) |
+            Type::Union(_) |
+            Type::Enumeration(_) |
+            Type::Bitfield(_) |
+            Type::Interface(_)
+            => {
+                let full_name = format!("{}.{}", &ns.name, t.get_name());
+                if env.type_status_sys(&full_name).ignored() {
+                    return None;
+                }
+                let name = match t.get_glib_name() {
+                    None => return None,
+                    Some(name) => name,
+                };
+                if is_name_made_up(&name) {
+                    return None;
+                }
+                let cfg_condition = env.config.objects.get(&full_name).and_then(|obj| {
+                    obj.cfg_condition.clone()
+                });
+                Some(CType {
+                    name: name.to_owned(),
+                    cfg_condition,
+                })
+            },
+            _ => None,
+        })
+        .collect::<Vec<_>>();
+
+    if ctypes.is_empty() {
+        return;
+    }
+
+    if !manual_h.exists() {
+        save_to_file(&manual_h, env.config.make_backup, |w| {
+            info!("Generating file {:?}", &manual_h);
+            writeln!(w, "// Insert manual customizations here, they won't be overwritten.")
+        });
+    }
+    save_to_file(&abi_c, env.config.make_backup, |w| {
+        generate_abi_c(env, &abi_c, w)
+    });
+    save_to_file(&abi_rs, env.config.make_backup, |w| {
+        generate_abi_rs(env, &abi_rs, w, &ctypes)
+    });
+}
+
+/// Checks if type name is unlikely to correspond to a real C type name.
+fn is_name_made_up(name: &str) -> bool {
+    // Unnamed types are assigned name during parsing, those names contain an underscore.
+    name.contains('_')
+}
+
+fn generate_abi_c(env: &Env, path: &Path, w: &mut Write) -> io::Result<()> {
+    info!("Generating file {:?}", path);
+    general::start_comments(w, &env.config)?;
```

Comment:
Please add `try!(writeln!(w, ""));` after this, and for `abi.rs` too.

Reply:
Done